In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
import pickle

## Processed Data

In [11]:
idx = 0
with open(f"2023-03-06_data/{idx}.degree.pkl", "rb") as f:
    dict = pickle.load(f)

In [8]:
len(dict['instance_idx'])

2950019

In [19]:
pd.Series(dict['net_degrees']).value_counts().sort_index()

0            857
1          15850
2         525902
3         130515
4          56196
           ...  
2050           2
2139           1
2141           1
4087           1
140605         1
Length: 304, dtype: int64

In [24]:
pd.Series(dict['cell_degrees']).value_counts().sort_index()

0          257
2        99955
3       324927
4       144497
5       217291
6         3999
7         6837
14           1
15           1
30           1
32           6
33           1
50           6
52           1
72           5
75           1
76           3
77           1
78           1
82           2
87           4
91           1
108         24
112          2
113          3
114          1
117         24
118         14
130          1
141          2
143         24
165          2
169          6
177         12
178          1
181          2
310          2
380          6
1243        14
dtype: int64

In [27]:
with open(f"2023-03-06_data/{idx}.bipartite.pkl", "rb") as f:
    dict = pickle.load(f)

In [28]:
dict

{'instance_idx': array([478481, 478482, 478483, ..., 797724, 738342, 797725]),
 'net_idx': array([     0,      0,      0, ..., 821521, 821522, 821522]),
 'edge_attr': array([3, 3, 3, ..., 2, 2, 2]),
 'edge_dir': array([0, 0, 0, ..., 1, 0, 1]),
 'sample_name': '/data/son/Research/chips/congestion_prediction/data/RosettaStone-GraphData-2023-03-06/superblue19/6/',
 'folder': '/data/son/Research/chips/congestion_prediction/data/RosettaStone-GraphData-2023-03-06/superblue1/1/',
 'design': 'superblue1',
 'edge_index': array([[     0,      0,      0, ..., 797724, 821522, 797725],
        [478481, 478482, 478483, ..., 821521, 738342, 821522]])}

In [29]:
with open(f"2023-03-06_data/{idx}.eigen.10.pkl", "rb") as f:
    dict = pickle.load(f)

In [32]:
dict['evals']

array([1.9982028, 1.9984511, 1.9999945, 1.9999961, 1.9999961, 1.999997 ,
       1.999999 , 1.9999994, 2.0000005, 2.0000098], dtype=float32)

In [33]:
dict['evects']

array([[ 1.73489370e-05,  1.81367905e-05, -4.42451652e-04, ...,
        -1.69613311e-04, -1.03775405e-04,  3.63318453e-04],
       [ 2.06509212e-05,  2.13837811e-05, -4.94658831e-04, ...,
        -1.89635786e-04, -1.15998235e-04,  4.06221749e-04],
       [ 2.06760851e-05,  2.14082029e-05, -4.94659995e-04, ...,
        -1.89635917e-04, -1.15999334e-04,  4.06223349e-04],
       ...,
       [-1.21696266e-05, -1.34866805e-05,  3.12872638e-04, ...,
         1.19940436e-04,  7.35051581e-05, -2.57036416e-04],
       [-1.18891930e-05, -1.27987933e-05,  3.12860706e-04, ...,
         1.19941426e-04,  7.34357745e-05, -2.56962288e-04],
       [-1.21403045e-05, -1.31762026e-05,  3.12862045e-04, ...,
         1.19949342e-04,  7.34591595e-05, -2.56991305e-04]], dtype=float32)

In [34]:
with open(f"2023-03-06_data/{idx}.global_information.pkl", "rb") as f:
    dict = pickle.load(f)

In [35]:
dict

{'design': 'superblue1',
 'variant': 1,
 'core_utilization': 0.5,
 'max_routing_layer': 'met5',
 'clk_per': 50,
 'clk_uncertainty': 0.2,
 'flow_stage': 'place_opt',
 'hstrap_layer': 'met5',
 'hstrap_width': 1.6,
 'hstrap_pitch': 27.2,
 'vstrap_layer': 'met4',
 'vstrap_width': 1.5,
 'vstrap_pitch': 27.3}

In [36]:
with open(f"2023-03-06_data/{idx}.metis_part_dict.pkl", "rb") as f:
    dict = pickle.load(f)

In [42]:
list(dict.keys())[:10]

[115, 181, 183, 184, 185, 186, 187, 247, 248, 250]

In [43]:
with open(f"2023-03-06_data/{idx}.net_demand_capacity.pkl", "rb") as f:
    dict = pickle.load(f)

In [48]:
pd.Series(dict['demand']).value_counts() #guessing power demand?

22.000000    8465
21.000000    8238
23.000000    8180
24.000000    7839
20.000000    7553
             ... 
23.381044       1
23.357000       1
22.473118       1
6.394209        1
51.155556       1
Length: 178589, dtype: int64

In [54]:
np.mean(dict['demand']), np.std(dict['demand'])

(21.723934436378457, 7.046322395240629)

In [55]:
with open(f"2023-03-06_data/{idx}.net_features.pkl", "rb") as f:
    dict = pickle.load(f)

In [59]:
len(dict['instance_features'])

821523

In [61]:
len(dict['instance_features'][0])

51

In [63]:
dict['instance_features'][0] #???

array([ 9.8392700e+05,  2.6645090e+02,  0.0000000e+00,  4.1767200e+05,
        1.4504403e+05,  3.1538994e+05,  1.4943203e+05,  3.9212781e+05,
        5.4454250e+05,  1.1225168e+05,  2.7461322e+05,  2.0919844e+05,
        1.7670420e+04,  5.0648119e+05,  8.4914600e+05,  4.0419009e+05,
        1.3343519e+06,  1.4035854e+06,  2.6564066e+05,  1.0652486e+06,
        6.0974419e+05,  4.1504445e+04,  1.2290500e+05,  2.6057900e+05,
        4.7308500e+05,  7.7726800e+05,  1.1596730e+06,  1.8579820e+06,
        4.2378000e+05,  7.0933500e+05,  1.1935630e+06,  1.8414130e+06,
        2.7894380e+06,  4.1105650e+06,  1.9671024e+00,  2.0726898e+00,
       -5.3799072e+01, -6.0643272e+01,  1.4420410e+00, -4.4606724e+00,
       -2.9280889e+00, -2.0650249e+01, -1.2616593e+01,  4.4177708e+01,
        1.7341040e+01,  4.1642575e+00,  5.9796691e-01,  7.7328318e-01,
        3.5756442e-01,  5.6296331e-01,  8.2648188e-01], dtype=float32)

In [64]:
with open(f"2023-03-06_data/{idx}.node_features.pkl", "rb") as f:
    dict = pickle.load(f)

In [65]:
dict

{'num_instances': 797938,
 'num_nets': 821523,
 'x_min': 1920,
 'x_max': 7278720,
 'y_min': 3330,
 'y_max': 7276050,
 'min_cell_width': 1920,
 'max_cell_width': 19211520,
 'min_cell_height': 13320,
 'max_cell_height': 20601600,
 'instance_features': array([[7.77638522e-01, 2.67399267e-01, 1.32000000e+02, 3.99800100e-04,
         0.00000000e+00, 6.00000000e+00],
        [6.15699208e-01, 4.91758242e-01, 3.59000000e+02, 7.99600200e-04,
         0.00000000e+00, 6.00000000e+00],
        [5.91820580e-01, 4.78021978e-01, 1.76000000e+02, 4.99750125e-04,
         0.00000000e+00, 0.00000000e+00],
        ...,
        [7.77506596e-01, 1.34615385e-01, 6.00000000e+00, 1.59920040e-03,
         0.00000000e+00, 6.00000000e+00],
        [7.75593668e-01, 9.34065934e-02, 6.00000000e+00, 1.59920040e-03,
         0.00000000e+00, 0.00000000e+00],
        [7.76781003e-01, 5.86080586e-02, 6.00000000e+00, 1.59920040e-03,
         0.00000000e+00, 0.00000000e+00]]),
 'sample_name': '/data/zluo/new_data/superblue

In [69]:
len(dict['instance_features'][0])

6

In [70]:
dict['instance_features'][0] #???

array([7.77638522e-01, 2.67399267e-01, 1.32000000e+02, 3.99800100e-04,
       0.00000000e+00, 6.00000000e+00])

In general, a bit confused about how the data is layout/partitioned. Is one index a netlist? Whole chip? Additionally, a lot of feature vectors that are not specified exactly what they are in the data writeup.


In [72]:
with open(f"2023-03-06_data/node_neighbors/{idx}.node_neighbor_features.pkl", "rb") as f:
    dict = pickle.load(f)

In [74]:
dict.keys()

dict_keys(['pd', 'neighbor'])

In [75]:
dict['pd']

array([[0.57659997, 3.78525   , 2.84100841, ..., 0.        , 0.        ,
        0.        ],
       [1.1236587 , 2.59016081, 0.48024168, ..., 0.        , 0.        ,
        0.        ],
       [1.1236587 , 2.59016081, 0.48024168, ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.95083978, 0.6301005 ,
        0.03481547],
       [0.        , 0.        , 0.        , ..., 1.14630555, 1.43355668,
        0.14265637],
       [0.        , 0.        , 0.        , ..., 1.44839475, 1.73426691,
        0.1504663 ]])

In [81]:
len(dict['pd'])

797938

In [86]:
len(dict['neighbor'][0]) # Again confused about what these vectors mean, and why do each of pd's and neighbor's elements differ in length

# ok pd is persistant diagram so makes sense for length to be diff

12

In [231]:
idx = 0
info = 'bipartite'
def open_data(idx, info):
    with open(f"2023-03-06_data/{idx}.{info}.pkl", "rb") as f:
        dict = pickle.load(f)
    return dict

### Bipartite

A bipartite representation of a chip, with wires/nets being represented as nodes on top of cells/instances and the pins/terminals being edges.

In [232]:
data = open_data(0, 'bipartite')

In [236]:
data

{'instance_idx': array([478481, 478482, 478483, ..., 797724, 738342, 797725]),
 'net_idx': array([     0,      0,      0, ..., 821521, 821522, 821522]),
 'edge_attr': array([3, 3, 3, ..., 2, 2, 2]),
 'edge_dir': array([0, 0, 0, ..., 1, 0, 1]),
 'sample_name': '/data/son/Research/chips/congestion_prediction/data/RosettaStone-GraphData-2023-03-06/superblue19/6/',
 'folder': '/data/son/Research/chips/congestion_prediction/data/RosettaStone-GraphData-2023-03-06/superblue1/1/',
 'design': 'superblue1',
 'edge_index': array([[     0,      0,      0, ..., 797724, 821522, 797725],
        [478481, 478482, 478483, ..., 821521, 738342, 821522]])}

In [235]:
data.keys()

dict_keys(['instance_idx', 'net_idx', 'edge_attr', 'edge_dir', 'sample_name', 'folder', 'design', 'edge_index'])

In [241]:
len(data['instance_idx']), len(data['net_idx'])

(2950019, 2950019)

'instance_idx', 'net_idx', 'edge_attr', 'edge_dir', and both elements in 'edge_index' have same length (295019)

In [256]:
len(set(data['instance_idx']))

797681

In [263]:
pd.Series(data['instance_idx']).value_counts().iloc[:20]

732495    1243
732475    1243
732516    1243
732509    1243
732484    1243
732505    1243
732479    1243
732514    1243
732490    1243
732507    1243
732481    1243
732519    1243
732497    1243
732503    1243
732458     380
732448     380
732442     380
732464     380
732453     380
732468     380
dtype: int64

In [289]:
pd.Series(pd.Series(data['instance_idx']).value_counts().values).value_counts()

3       324975
5       216932
4       144290
2        99906
7         7016
6         4387
118         38
109         24
144         24
1243        14
178         12
169          6
380          6
32           6
50           6
72           5
87           4
113          3
76           3
142          2
310          2
181          2
82           2
165          2
112          2
75           1
53           1
77           1
33           1
78           1
30           1
15           1
14           1
91           1
179          1
114          1
130          1
dtype: int64

Many duplicate indices within 'instance_idx', not exactly sure where these indices coming from.

Not from cell library data as only 80 different cells within that.

Most indices appear a few times (2-7), though there are certain with specific high frequency (ex. 118, 169, 380, 1243).

No index appears only once.

In [310]:
pd.Series((data['edge_attr'])).value_counts()

1       797681
2       797681
3       697775
4       372800
5       228510
         ...  
671         14
672         14
673         14
674         14
1243        14
Length: 1243, dtype: int64

797681 matches up with the unique number of instance indices.

Makes sense as each instance has at least one pin and this pin must be used.

Tells us each instance has at least 2 pins being used.

In [311]:
len(data['edge_attr'])

2950019

I am assuming edge_attr is the pin/terminal index that is being used by the net/wire to connect to the cell.

In [308]:
data['edge_dir']

array([0, 0, 0, ..., 1, 0, 1])

'edge_dir' only consists of 0s and 1s, assuming signifies which direction the edge is pointing (not exactly sure which direction is for 0 and which for 1).

In [254]:
len(data['edge_index'][1])

2950019

In [291]:
data.keys()

dict_keys(['instance_idx', 'net_idx', 'edge_attr', 'edge_dir', 'sample_name', 'folder', 'design', 'edge_index'])

In [294]:
def get_elem_data(idx):
    print(data['instance_idx'][idx])
    print(data['net_idx'][idx])
    print(data['edge_attr'][idx])
    print(data['edge_dir'][idx])
    print(data['edge_index'][0][idx])
    print(data['edge_index'][1][idx])

In [306]:
get_elem_data(12030)

492078
0
3
0
0
492078


## Raw Data


In [156]:
import gzip
import json

In [157]:
with gzip.open('NCSU-DigIC-GraphData-2023-07-25/cells.json.gz', 'rt') as file:
    cell_data = json.load(file)

In [158]:
len(cell_data)

80

In [174]:
cell_data[0]

{'name': 'AND2_X1',
 'width': 768,
 'height': 1536,
 'terms': [{'name': 'A1', 'id': 1, 'dir': 0},
  {'name': 'A2', 'id': 2, 'dir': 0},
  {'name': 'Z', 'id': 3, 'dir': 1}],
 'id': 0}

cells.json.gz stores cell library data.

'terms' are the pins of the cell.

In [186]:
cell_data[1]['terms']

[{'name': 'A1', 'id': 1, 'dir': 0},
 {'name': 'A2', 'id': 2, 'dir': 0},
 {'name': 'Z', 'id': 3, 'dir': 1}]

- 'id' is re

### Counter

#### Features

In [160]:
with gzip.open('NCSU-DigIC-GraphData-2023-07-25/counter/1/counter.json.gz', 'rt') as file:
    counter_feature_data = json.load(file)

In [161]:
len(counter_feature_data['instances']), len(counter_feature_data['nets'])

(23, 28)

23 instances and 28 nets in counter

Instances are cells with pins and nets are wires

In [162]:
counter_feature_data['instances'][0]

{'name': 'clk_gate_value_reg/latch',
 'id': 0,
 'xloc': 2560,
 'yloc': 7680,
 'cell': 23,
 'orient': 0}

- name(str): name of the instance
- id(int): index of the instance in the array
    - index just in this chip
- xloc, yloc(int): location of the instance in database units (DBU)
- cell(int): master library cell ID (index)
    - what kind of cell that is being instantiated
- orient(int): number from 0 to 7 for orientation of instance

In [168]:
pd.Series([x['cell'] for x in counter_feature_data['instances']]).value_counts()

17    5
65    3
48    3
7     3
34    2
20    2
23    1
58    1
42    1
21    1
15    1
dtype: int64

different types of cells in counter

In [164]:
[x for x in counter_feature_data['instances'] if x['cell'] == 17]

[{'name': 'ropt_h_inst_941',
  'id': 14,
  'xloc': 4608,
  'yloc': 7680,
  'cell': 17,
  'orient': 6},
 {'name': 'copt_h_inst_791',
  'id': 16,
  'xloc': 4864,
  'yloc': 7680,
  'cell': 17,
  'orient': 0},
 {'name': 'copt_h_inst_793',
  'id': 17,
  'xloc': 7552,
  'yloc': 7680,
  'cell': 17,
  'orient': 6},
 {'name': 'copt_h_inst_796',
  'id': 20,
  'xloc': 4992,
  'yloc': 4608,
  'cell': 17,
  'orient': 0},
 {'name': 'ropt_h_inst_943',
  'id': 22,
  'xloc': 6912,
  'yloc': 4608,
  'cell': 17,
  'orient': 0}]

In [165]:
counter_feature_data['nets'][0]

{'name': 'clock', 'id': 0}

In [166]:
[x['name'] for x in counter_feature_data['nets']]

['clock',
 'reset',
 'dec',
 'zero',
 'test_si',
 'test_so',
 'test_se',
 'copt_net_53',
 'ropt_net_56',
 'aps_rename_1_',
 'value_1_',
 'value_0_',
 'N9',
 'net41',
 'n12',
 'n13',
 'n14',
 'n15',
 'n18',
 'n19',
 'n20',
 'n23',
 'copt_net_49',
 'copt_net_50',
 'copt_net_52',
 'copt_net_55',
 'ropt_net_57',
 'ropt_net_58']

#### Connectivity/Incidence Matrix


In [189]:
data = np.load('NCSU-DigIC-GraphData-2023-07-25/counter/1/counter_connectivity.npz')

In [190]:
data.files

['row', 'col', 'data', 'shape']

In [116]:
data['row']

array([ 0,  7,  9, 10, 13, 13,  6, 11,  5,  1,  6, 19,  1,  2,  3,  0,  2,
        8,  4, 18, 15, 14,  5, 21,  2, 22,  3, 14,  1, 12,  0,  1,  2,  3,
        0,  7,  6,  5,  6,  9,  4,  9,  8, 13, 11,  3,  9,  2,  7,  1, 10,
       12, 13, 16, 15, 20, 16, 18, 17,  3, 10,  8,  4, 20, 19, 21, 17, 22])

In [117]:
data['col']

array([ 0,  1,  1,  1,  1,  2,  3,  3,  3,  4,  5,  5,  6,  6,  6,  6,  7,
        7,  7,  7,  8,  8,  9,  9,  9, 10, 10, 11, 11, 12, 12, 13, 13, 13,
       13, 14, 14, 15, 15, 15, 15, 16, 16, 17, 17, 18, 18, 19, 19, 20, 20,
       21, 21, 22, 22, 23, 23, 24, 24, 25, 25, 25, 25, 25, 26, 26, 27, 27])

In [118]:
data['data']

array([1, 1, 3, 1, 3, 1, 3, 1, 3, 3, 1, 2, 2, 2, 2, 3, 3, 2, 1, 2, 1, 2,
       1, 1, 6, 1, 6, 1, 6, 2, 2, 5, 5, 5, 4, 2, 4, 2, 2, 2, 3, 1, 3, 2,
       2, 1, 4, 1, 3, 1, 3, 1, 4, 1, 2, 1, 2, 1, 2, 3, 2, 1, 2, 2, 1, 2,
       1, 2])

In [119]:
data['shape']

array([23, 28])

In [191]:
from scipy.sparse import coo_matrix
coo = coo_matrix((data['data'], (data['row'], data['col'])), shape=data['shape'])

In [194]:
coo.toarray()[0]

array([1, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 2, 4, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0])

Signifies that in this instance/cell (index 0 instance), they are connected to index 0 net/wire at the instance's first pin/terminal. They are also connected to the index 6 net/wire at the instance's 3rd pin. 

The pins indicies just based on order they are (placed) in the instance. Can access pin information from the 'terms' in the instance's cell data (from cells.json.gz).

In [169]:
counter_congestion_data = np.load('NCSU-DigIC-GraphData-2023-07-25/counter/1/counter_congestion.npz')

In [170]:
counter_congestion_data.files

['demand', 'capacity', 'xBoundaryList', 'yBoundaryList', 'layerList']

In [221]:
print(counter_congestion_data['layerList'])

['M1' 'MINT1' 'MINT2' 'MINT3' 'MINT4' 'MINT5' 'MSMG1' 'MSMG2' 'MSMG3'
 'MSMG4' 'MSMG5' 'MG1' 'MG2']


In [203]:
ybl = counter_congestion_data['yBoundaryList']
xbl = counter_congestion_data['xBoundaryList']
xbl, ybl

(array([   0, 2048, 3584, 5120, 6656]),
 array([   0, 1536, 3072, 4608, 6144, 7680, 9216]))

In [206]:
def getGRCIndex(x,y,xbl,ybl):
    j=0
    for b in xbl[1:]:
        if x<b:break
        j+=1
    i=0
    for b in ybl[1:]:
        if y<b:break
        i+=1
    return i,j

In [218]:
xloc = 5048
yloc = 3536

In [219]:
getGRCIndex(xloc, yloc, xbl, ybl)

(2, 2)

Global route cell (GRC) is a higher-level representation of a group of cells, most likely partitioned by purpose and physical location.

Used early on in routing process of chip design to predict areas of higher congestion.

In [313]:
def buildBST(array,start=0,finish=-1):
    if finish<0:
        finish = len(array)
    mid = (start + finish) // 2
    if mid-start==1:
        ltl=start
    else:
        ltl=buildBST(array,start,mid)
    if finish-mid==1:
        gtl=mid
    else:
        gtl=buildBST(array,mid,finish)
    return((array[mid],ltl,gtl))
xbst=buildBST(counter_congestion_data['xBoundaryList'])
ybst=buildBST(counter_congestion_data['yBoundaryList'])

In [314]:
def getGRCIndex(x, y, xbst, ybst):
    while (type(xbst)==tuple):
        if x < xbst[0]:
            xbst=xbst[1]
        else:
            xbst=xbst[2]

    while (type(ybst)==tuple):
        if y < ybst[0]:
            ybst=ybst[1]
        else:
            ybst=ybst[2]
    return ybst, xbst

In [319]:
getGRCIndex(5000, 3000, xbst, ybst)

(1, 2)

In [172]:
counter_congestion_data['demand'][3]

array([[0, 1, 1, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 1],
       [0, 0, 1, 0, 0],
       [1, 2, 1, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0]])

In [222]:
counter_congestion_data['capacity'][3]

array([[11, 11, 11, 11, 11],
       [12, 12, 12, 12, 12],
       [10, 11, 12, 12, 12],
       [11, 12, 12, 12, 12],
       [11, 11, 12, 12, 12],
       [12, 12, 12, 12, 12],
       [11, 11, 12, 12, 12]])

- Demand is the number of wires required to pass through the certain GRC
- Capacity is the number of wires that can pass through without congestion

Each element in the matrix refers to a subsection of the GRC.

How are subsections partitioned? Why are all the GRCs the same shape in counter?

Is it possible different GRCs actually have different sizes, scaled by x and y boundaries. Just arbitrarily separated into 7 by 5 matrices?

In [228]:
counter_congestion_data['demand'][10].shape

(7, 5)

In [323]:
counter_feature_data['instances']

[{'name': 'clk_gate_value_reg/latch',
  'id': 0,
  'xloc': 2560,
  'yloc': 7680,
  'cell': 23,
  'orient': 0},
 {'name': 'value_reg_0_',
  'id': 1,
  'xloc': 512,
  'yloc': 4608,
  'cell': 65,
  'orient': 0},
 {'name': 'value_reg_2_',
  'id': 2,
  'xloc': 768,
  'yloc': 7680,
  'cell': 65,
  'orient': 6},
 {'name': 'value_reg_1_',
  'id': 3,
  'xloc': 3584,
  'yloc': 4608,
  'cell': 65,
  'orient': 6},
 {'name': 'U20', 'id': 4, 'xloc': 7424, 'yloc': 4608, 'cell': 58, 'orient': 6},
 {'name': 'U21', 'id': 5, 'xloc': 6272, 'yloc': 7680, 'cell': 48, 'orient': 6},
 {'name': 'U22', 'id': 6, 'xloc': 6784, 'yloc': 7680, 'cell': 7, 'orient': 6},
 {'name': 'U23', 'id': 7, 'xloc': 5760, 'yloc': 7680, 'cell': 48, 'orient': 2},
 {'name': 'U24', 'id': 8, 'xloc': 6912, 'yloc': 4608, 'cell': 42, 'orient': 4},
 {'name': 'U25', 'id': 9, 'xloc': 6400, 'yloc': 4608, 'cell': 7, 'orient': 4},
 {'name': 'U26',
  'id': 10,
  'xloc': 6272,
  'yloc': 7680,
  'cell': 48,
  'orient': 2},
 {'name': 'U28',
  'id': 

In [325]:
cell_data[0]

{'name': 'AND2_X1',
 'width': 768,
 'height': 1536,
 'terms': [{'name': 'A1', 'id': 1, 'dir': 0},
  {'name': 'A2', 'id': 2, 'dir': 0},
  {'name': 'Z', 'id': 3, 'dir': 1}],
 'id': 0}